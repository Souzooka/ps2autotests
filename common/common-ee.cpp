#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
