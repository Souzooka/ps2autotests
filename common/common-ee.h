#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
