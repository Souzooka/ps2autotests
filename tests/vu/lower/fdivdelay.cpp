#include <common-ee.h>
#include "../test_runner.h"

static VU::LowerOp DIV_adapter(VU::Field field, VU::Reg reg) {
	using namespace VU;
	return DIV(FIELD_W, VF00, field, reg);
}

static VU::LowerOp RSQRT_adapter(VU::Field field, VU::Reg reg) {
	using namespace VU;
	return RSQRT(FIELD_W, VF00, field, reg);
}

class FdivDelayTestRunner : public TestRunner {
public:
	FdivDelayTestRunner(int vu) : TestRunner(vu) {
	}

	typedef VU::LowerOp (*FdivFunction)(VU::Field field, VU::Reg reg);

	void PerformSingleWaitQ(const char *name, FdivFunction func) {
		using namespace VU;

		printf("%s single waitq:\n", name);

		Reset();

		//Prepare registers
		WrImm(4.0f);
		Wr(MULi(DEST_W, VF01, VF00));
		
		//Op
		Wr(func(FIELD_W, VF01));
		Wr(WAITQ());
		Wr(MULq(DEST_XYZW, VF01, VF00));
		
		Execute();

		printf("  VF01: "); PrintRegister(VF01, true);
	}

	void PerformMultipleWaitQ(const char *name, FdivFunction func) {
		using namespace VU;

		printf("%s multiple waitq:\n", name);

		Reset();

		//Prepare registers
		WrImm(4.0f);
		WrImm(MULi(DEST_W, VF01, VF00), 16.0f);
		Wr(MULi(DEST_W, VF02, VF00));
		
		//Op
		Wr(func(FIELD_W, VF01));
		Wr(WAITQ());
		Wr(MULq(DEST_XYZW, VF01, VF00));
		Wr(func(FIELD_W, VF02));
		Wr(WAITQ());
		Wr(MULq(DEST_XYZW, VF02, VF00));

		Execute();

		printf("  VF01: "); PrintRegister(VF01, true);
		printf("  VF02: "); PrintRegister(VF02, true);
	}

	void PerformSingleDelayed(const char *name, FdivFunction func) {
		using namespace VU;

		printf("%s single delayed:\n", name);

		Reset();

		//Set Q to 1.0f
		Wr(DIV(FIELD_W, VF00, FIELD_W, VF00));
		Wr(WAITQ());

		//Prepare registers
		WrImm(4.0f);
		Wr(MULi(DEST_W, VF01, VF00));
		
		//Op
		Wr(func(FIELD_W, VF01));
		Wr(MULq(DEST_XYZW, VF01, VF00));
		Wr(MULq(DEST_XYZW, VF02, VF00));
		Wr(MULq(DEST_XYZW, VF03, VF00));
		Wr(MULq(DEST_XYZW, VF04, VF00));
		Wr(MULq(DEST_XYZW, VF05, VF00));
		Wr(MULq(DEST_XYZW, VF06, VF00));
		Wr(MULq(DEST_XYZW, VF07, VF00));
		Wr(MULq(DEST_XYZW, VF08, VF00));
		Wr(MULq(DEST_XYZW, VF09, VF00));
		Wr(MULq(DEST_XYZW, VF10, VF00));
		Wr(MULq(DEST_XYZW, VF11, VF00));
		Wr(MULq(DEST_XYZW, VF12, VF00));
		Wr(MULq(DEST_XYZW, VF13, VF00));
		Wr(MULq(DEST_XYZW, VF14, VF00));
		
		Execute();
		
		for(u32 i = 1; i <= 14; i++) {
			printf("  VF%02d: ", i); PrintRegister(static_cast<VU::Reg>(VF00 + i), true);
		}
	}
	
	void PerformMultipleDelayed1(const char *name, FdivFunction func) {
		using namespace VU;

		printf("%s multiple delayed (1):\n", name);

		Reset();

		//Set Q to 1.0f
		Wr(DIV(FIELD_W, VF00, FIELD_W, VF00));
		Wr(WAITQ());

		//Prepare registers
		WrImm(4.0f);
		WrImm(MULi(DEST_W, VF01, VF00), 16.0f);
		Wr(MULi(DEST_W, VF02, VF00));
		
		//Op
		Wr(func(FIELD_W, VF01));
		Wr(func(FIELD_W, VF02));
		Wr(MULq(DEST_XYZW, VF01, VF00));
		Wr(MULq(DEST_XYZW, VF02, VF00));
		Wr(MULq(DEST_XYZW, VF03, VF00));
		Wr(MULq(DEST_XYZW, VF04, VF00));
		Wr(MULq(DEST_XYZW, VF05, VF00));
		Wr(MULq(DEST_XYZW, VF06, VF00));
		Wr(MULq(DEST_XYZW, VF07, VF00));
		Wr(MULq(DEST_XYZW, VF08, VF00));
		Wr(MULq(DEST_XYZW, VF09, VF00));
		Wr(MULq(DEST_XYZW, VF10, VF00));
		Wr(MULq(DEST_XYZW, VF11, VF00));
		Wr(MULq(DEST_XYZW, VF12, VF00));
		Wr(MULq(DEST_XYZW, VF13, VF00));
		Wr(MULq(DEST_XYZW, VF14, VF00));
		
		Execute();
		
		for(u32 i = 1; i <= 14; i++) {
			printf("  VF%02d: ", i); PrintRegister(static_cast<VU::Reg>(VF00 + i), true);
		}
	}
	
	void PerformMultipleDelayed2(const char *name, FdivFunction func) {
		using namespace VU;

		printf("%s multiple delayed (2):\n", name);

		Reset();

		//Set Q to 1.0f
		Wr(DIV(FIELD_W, VF00, FIELD_W, VF00));
		Wr(WAITQ());

		//Prepare registers
		WrImm(4.0f);
		WrImm(MULi(DEST_W, VF01, VF00), 16.0f);
		Wr(MULi(DEST_W, VF02, VF00));
		
		//Op
		Wr(func(FIELD_W, VF01));
		Wr(MULq(DEST_XYZW, VF01, VF00));
		Wr(func(FIELD_W, VF02));
		Wr(MULq(DEST_XYZW, VF02, VF00));
		Wr(MULq(DEST_XYZW, VF03, VF00));
		Wr(MULq(DEST_XYZW, VF04, VF00));
		Wr(MULq(DEST_XYZW, VF05, VF00));
		Wr(MULq(DEST_XYZW, VF06, VF00));
		Wr(MULq(DEST_XYZW, VF07, VF00));
		Wr(MULq(DEST_XYZW, VF08, VF00));
		Wr(MULq(DEST_XYZW, VF09, VF00));
		Wr(MULq(DEST_XYZW, VF10, VF00));
		Wr(MULq(DEST_XYZW, VF11, VF00));
		Wr(MULq(DEST_XYZW, VF12, VF00));
		Wr(MULq(DEST_XYZW, VF13, VF00));
		Wr(MULq(DEST_XYZW, VF14, VF00));
		
		Execute();
		
		for(u32 i = 1; i <= 14; i++) {
			printf("  VF%02d: ", i); PrintRegister(static_cast<VU::Reg>(VF00 + i), true);
		}
	}

	void Perform(const char *name, FdivFunction func) {
		using namespace VU;
		PerformSingleWaitQ(name, func);
		PerformMultipleWaitQ(name, func);
		PerformSingleDelayed(name, func);
		PerformMultipleDelayed1(name, func);
		PerformMultipleDelayed2(name, func);
	}
};

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	FdivDelayTestRunner runner(0);

	runner.Perform("DIV",   &DIV_adapter);
	runner.Perform("SQRT",  &VU::SQRT);
	runner.Perform("RSQRT", &RSQRT_adapter);
	
	printf("-- TEST END\n");
	return 0;
}
