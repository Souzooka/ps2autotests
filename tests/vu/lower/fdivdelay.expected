-- TEST BEGIN
DIV single waitq:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
DIV multiple waitq:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
DIV single delayed:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
DIV multiple delayed (1):
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
DIV multiple delayed (2):
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3d800000/+0.06
SQRT single waitq:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
SQRT multiple waitq:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
SQRT single delayed:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
SQRT multiple delayed (1):
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
SQRT multiple delayed (2):
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40000000/+2.00
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 40800000/+4.00
RSQRT single waitq:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
RSQRT multiple waitq:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
RSQRT single delayed:
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
RSQRT multiple delayed (1):
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
RSQRT multiple delayed (2):
  VF01: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f800000/+1.00
  VF02: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF03: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF04: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF05: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF06: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF07: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF08: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF09: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF10: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF11: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF12: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF13: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3f000000/+0.50
  VF14: 00000000/+0.00 00000000/+0.00 00000000/+0.00 3e800000/+0.25
-- TEST END
